pkg/cmd/server/server.go

```diff
@@ -0,0 +1,172 @@
+// of this server will be able to use this options object as a sub
+// options of it's own.
+type ServerOptions struct {
+	// the runtime configuration of our server
+	GenericServerRunOptions *genericserveroptions.ServerRunOptions
+	// the https configuration. certs, etc
+	SecureServingOptions *genericserveroptions.SecureServingOptions
+	// storage with etcd
+	EtcdOptions *genericserveroptions.EtcdOptions
+	// authn
+	AuthenticationOptions *genericserveroptions.DelegatingAuthenticationOptions
+	// authz
+	AuthorizationOptions *genericserveroptions.DelegatingAuthorizationOptions
+}
+
+// I made this up to match some existing paths. I am not sure if there
+// are any restrictions on the format or structure beyond text
+// separated by slashes.
+const etcdPathPrefix = "/k8s.io/incubator/service-catalog"
+
+// I made this up. Maybe we'll need it.
+const GroupName = "service-catalog.incubator.k8s.io"
+
+func NewCommandServer(out io.Writer) *cobra.Command {
+	// initalize our sub options.
+	options := &ServerOptions{
+		GenericServerRunOptions: genericserveroptions.NewServerRunOptions(),
+		SecureServingOptions:    genericserveroptions.NewSecureServingOptions(),
+		EtcdOptions:             genericserveroptions.NewEtcdOptions(),
+		AuthenticationOptions:   genericserveroptions.NewDelegatingAuthenticationOptions(),
+		AuthorizationOptions:    genericserveroptions.NewDelegatingAuthorizationOptions(),
+	}
+
+	// when our etcd resources are created, put them in a location
+	// specific to us
+	options.EtcdOptions.StorageConfig.Prefix = etcdPathPrefix
+
+	// I have no idea what this line does but I keep seeing it. Do
+	// we have our own `GroupName`? Is it like the pathPrefix for
+	// etcd?
+	//
+	// options.EtcdOptions.StorageConfig.Codec = api.Codecs.LegacyCodec(registered.EnabledVersionsForGroup(api.GroupName)...)
+
+	// this is the one thing that this program does. It runs the apiserver.
+	cmd := &cobra.Command{
+		Short: "run a service-catalog server",
+		Run: func(c *cobra.Command, args []string) {
+			options.runServer()
+		},
+	}
+
+	// We pass flags object to sub option structs to have them configure
+	// themselves. Each options adds it's own command line flags
+	// in addition to the flags that are defined above.
+	flags := cmd.Flags()
+	// TODO consider an AddFlags() method on our options
+	// struct. Will need to import pflag.
+	//
+	// repeated pattern seems like it should be refactored if all
+	// options were of an interface type that specified AddFlags.
+	options.GenericServerRunOptions.AddUniversalFlags(flags)
+	options.SecureServingOptions.AddFlags(flags)
+	options.EtcdOptions.AddFlags(flags)
+	options.AuthenticationOptions.AddFlags(flags)
+	options.AuthorizationOptions.AddFlags(flags)
+
+	return cmd
+}
+
+// runServer is a method on the options for composition. allows embedding in a higher level options as we do the etcd and serving options.
+func (serverOptions ServerOptions) runServer() error {
+	fmt.Println("set up the server")
```

Comment:
I agree.

I know almost nothing about logging in golang/k8s. Do we have a project-wide/k8s-wide doc or decision that I can refer to?